### PR TITLE
Add PHP types to private methods and functions

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Ref #47551, continues #49342
| License       | MIT
| Doc PR        | -

This PR adds real PHP types to private, internal or final (excluding `@final`) methods based on Psalm's type inference. Many more return types are missing still, but let's do them in other contributions.

This change can result in BC breaks in multiple ways, so we should carefully review the changes:

* A method may wrongly be considered "safe" by Psalter. If the method is unsafe (e.g. `@final`, but not yet real `final`), we should instead add a PHPdoc return type (see #49349)
* Psalm may not infer the correct set of return types. If an unexpected type is returned, this will result in a PHP error.

### CODE REVIEW

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Configuration.php [hunk 1]
```diff
@@ -290,47 +279,37 @@ public function writeBaseline()
...
-     * @return bool
-     */
-    public function verboseOutput($group)
+    public function verboseOutput($group): bool
     {
         return $this->verboseOutput[$group];
     }
 
-    public function shouldWriteToLogFile()
+    public function shouldWriteToLogFile(): bool: bool: bool: bool
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Configuration.php]: boo

File: src/Symfony/Component/Form/Console/Descriptor/JsonDescriptor.php [hunk 2]
```diff
@@ -33,7 +33,7 @@ protected function describeDefaults(array $options)
         $this->writeData($data, $options);
     }
 
-    protected function describeResolvedFormType(ResolvedFormTypeInterface $resolvedFormType, array $options = [])
+    prot: void: void: void: voidected function describeResolvedFormType(ResolvedFormTypeInterface $resolvedFormType, array $options = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Console/Descriptor/JsonDescriptor.php]: bug

File: src/Symfony/Component/Form/Console/Descriptor/JsonDescriptor.php [hunk 3]
```diff
@@ -61,7 +61,7 @@ protected function describeResolvedFormType(ResolvedFormTypeInterface $resolvedF
         $this->writeData($data, $options);
     }
 
-    protected function describeOption(OptionsResolver $optionsResolver, array $options)
+    prot: void: void: void: voidected function describeOption(OptionsResolver $optionsResolver, array $options)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Console/Descriptor/JsonDescriptor.php]: too

File: src/Symfony/Component/HttpKernel/Debug/FileLinkFormatter.php [hunk 4]
```diff
@@ -49,6 +49,11 @@ public function __construct(string|array $fileLinkFormat = null, RequestStack $r
         $this->urlFormat = $urlFormat;
     }
 
+    /**
+     * @return string|false
+     *
+     * @psalm-return false|string
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Debug/FileLinkFormatter.php]: needless

File: src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorFromClassMetadata.php [hunk 5]
```diff
@@ -65,7 +65,7 @@ public function getTypeForMappedObject(object|string $object): ?string
         return $mapping->getMappedObjectType($object);
     }
 
-    private function resolveMappingForMappedObject(object|string $object)
+    private function resolveMappingForMappedObject(object|string $object): ClassDiscriminatorMapping|null
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorFromClassMetadata.php]: `?ClassDiscriminatorMapping`

File: src/Symfony/Component/Workflow/Dumper/MermaidDumper.php [hunk 6]
```diff
@@ -231,7 +231,7 @@ private function styleWorkflowTransition(
         int $transitionId,
         string $transitionLabel,
         array $transitionMeta
-    ) {
+    ): array {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Workflow/Dumper/MermaidDumper.php]: I'd be in favor of putting method definitions on one line as we do elsewhere

File: src/Symfony/Component/Cache/Traits/RedisClusterNodeProxy.php [hunk 7]
```diff
@@ -35,12 +35,12 @@ public function __call(string $method, array $args)
...
     }
 
-    public function scan(&$iIterator, $strPattern = null, $iCount = null)
+    public function scan(&$iIterator, $strPattern = null, $iCount = null): array|bool
     {
         return $this->redis->scan($iIterator, $this->host, $strPattern, $iCount);
     }
 
-    public function getOption($name)
+    public function getOption($name): int
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisClusterNodeProxy.php]: better revert the changes on this file IMHO

File: src/Symfony/Component/VarExporter/Internal/Registry.php [hunk 8]
```diff
@@ -56,14 +56,14 @@ public static function p($class)
...
-    public static function f($class)
+    public static function f($class): callable
     {
         $reflector = self::$reflectors[$class] ??= self::getClassReflector($class, true, false);
 
         return self::$factories[$class] = [$reflector, 'newInstanceWithoutConstructor'](...);
     }
 
-    public static function getClassReflector($class, $instantiableWithoutConstructor = false, $cloneable = null)
+    public static function getClassReflector($class, $instantiableWithoutConstructor = false, $cloneable = null): \ReflectionClass
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarExporter/Internal/Registry.php]: `VarExporter\Internal` doesn't use type hint for perf

File: src/Symfony/Component/HttpKernel/Debug/FileLinkFormatter.php (hunk 4 above)
Comment (wouterj) [reply src/Symfony/Component/HttpKernel/Debug/FileLinkFormatter.php]: Yeah, we should remove all `@psalm-return` added, they were a mistake.

**Comment by nicolas-grekas** [comment]
Thank you @wouterj.
